File: test/corpus/propagation/struct_literal_fields.exs
defmodule RustQ.Corpus.Propagation.StructLiteralFields do
  @moduledoc "Struct literal fields are checked against expected field types."

  use RustQ.Meta

  alias RustQ.Type, as: R

  defmodule Point do
    defstruct [:x, :y]
  end

  @type point :: %Point{x: R.f32(), y: R.f32()}

  @spec decode_x(term()) :: R.nif_result(R.f32())
  defrust decode_x(term) do
    decode_as!(term, R.f32())
  end

  @spec consume(point()) :: R.nif_result(R.unit())
  defrust consume(_point) do
    :ok
  end

  @spec run(term()) :: R.nif_result(R.unit())
  defrust run(term) do
    point = struct_literal(Point, x: decode_x(term), y: 0.0)
    consume(point)
    :ok
  end
end

File: test/corpus/propagation/struct_literal_fields.rs
#[derive(Clone, Debug)]
pub struct Point {
    pub x: f32,
    pub y: f32,
}

pub fn decode_point<'a>(term: Term<'a>) -> NifResult<Point> {
    Ok(Point {
        x: term.map_get(atoms::x())?.decode()?,
        y: term.map_get(atoms::y())?.decode()?,
    })
}

fn decode_x<'a>(term: Term<'a>) -> NifResult<f32> {
    term.decode::<f32>()?
}

fn consume(_point: Point) -> NifResult<()> {
    Ok(())
}

fn run<'a>(term: Term<'a>) -> NifResult<()> {
    let point = Point {
        x: decode_x(term)?,
        y: 0.0,
    };
    consume(point)?;
    Ok(())
}
